[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161383
* __->__ #160900



convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.
@exported-using-ghexport

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

zhxchen17 committed:
[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]

convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -840,12 +840,122 @@ class DynamoOutput:
+    closure: tuple[CellType]
+
+
+def fullgraph_capture(frame: FrameInfo) -> CaptureOutput:
```
This seems like new functionality, and not used anywhere in the program. Can you add test for it?

anijain2305 replied:
Actually, is this what we want? From what I understand, you will use this for strict export, but there is no need to run the eval frame. `DynamoTracerOutput` should essentially everything you need for strict export.

zhxchen17 replied:
Also fullgraph_capture shouldn't run eval_frame, it expects a FrameInfo struct as input and this frame object can be constructed without eval_frame.

zhxchen17 replied:
> I am fine with calling eval_frame in this PR, but we should strive (soonish) to not rely on eval frame. This will be a very clean solution.

Sorry what do you mean by this? iiuc this pr doesnt use eval frame at all.

anijain2305 replied:
> Sorry what do you mean by this? iiuc this pr doesnt use eval frame at all.

Ok I just checked that fullgraph_capture takes a Frame. Not a requirement for this PR, but consider adding a test.

tugsbayasgalan replied:
For pytree information, we actually don't need the bytecode as we can setup our tracing function in following fashion:
```
def fn_to_trace(*flat_args):
      nonlocal input_spec 
      nonlocal out_spec 
      real_args, real_kwargs = pytree.tree_unflatten(flat_args, input_spec) 
      out  = fn(*real_args, **real_kwargs)
      out_flat, out_spec = pytree.tree_flatten(out)
      return out_flat
 ```
 
 Note that we can still trace this callable via dynamo to get the graph module. It is actually probably better to use dynamo here because:
 1) We can use VariableTracker to get the information about input/output relatively easily since they are all flat.
 2) There could be silent incorrectness in user defined pytree flatten/unflatten.  
 
 But I think we do still need bytecode for cases like:
 1) side effects 
 2) ...

zhxchen17 replied:
talked offline with @tugsbayasgalan, seems now we just want to wrap over the user function (eager transformation) and let dynamo trace the wrapped funciton. I'm fine with this approach and seems in the new plan there's no need to change the current fullgraph_capture code.

The old plan was to use bytecode in a export as a pass, but this requires rerun the model twice so I think the new plan seems cleaner to me.

Still precompile requires bytecode from DynamoOutput, so I think export should just ignore the bytecode produced here.

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"


convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 committed:
Update on "[dynamo] Refactor convert_frame.compile_frame to be self contained function. [5/n]"




convert_frame.compile_frame used to take a callback transform function which will capture the frame object it has, but the frame information is not passed directly into compile_frame function.

This PR changes the signature of compile_frame so that frame information is directly passed in the function without taking a callback. This makes it easier to build fullgraph capture API on top of compile_frame.
exported-using-ghexport

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801/)

Differential Revision: [D80469801](https://our.internmc.facebook.com/intern/diff/D80469801)

[ghstack-poisoned]

zhxchen17 commented:
added unit test and docstring for functions and data structures.

zhxchen17 commented:
@pytorchbot merge -i

atalman commented:
@pytorchmergebot revert -c ghfirst -m "reverting since can't land existing diff internally, will need to reland it"
